blog: using webhook with node-red (#1917)
FlowFuse/website: src/blog/2024/04/using-webhook-with-node-red.md

---
title: Using Webhooks with Node-RED
subtitle: Learn to Integrate and Automate with Webhooks and Node-RED.
description: Learn how to seamlessly integrate webhooks into your Node-RED applications for automating tasks and enhancing communication. Get detailed, step-by-step instructions on setting up webhooks and practical implementation examples.

> date: 2024-04-07

Reviewer: Need to pick a date.

authors: ["sumit-shinde"]
image:
tags:
    - post
    - node-red
    - webhooks
    - event-driven
    - automation
---

Have you found yourself needing to automate tasks based on specific events within your IoT applications? If so, utilizing webhooks presents an efficient solution for this challenge. Webhooks play an integral role in streamlining operations, facilitating real-time communication and seamless integration between disparate systems and devices. In this guide, we cover everything you need to get started with webhooks, from basics to practical implementation. We provide detailed, step-by-step instructions on integrating systems through webhooks using Node-RED.

<!--more-->

> ## What are the Webhooks?

Reviewer: ```suggestion
## What are the Webhooks?
```
```suggestion
## What are Webhooks?
```

Webhooks are a mechanism that allows fast communication between two different applications or services. They are essentially HTTP requests triggered by events in a source system and sent to a destination system, often with a payload of data. Webhooks are automated, meaning they are automatically sent out when their corresponding event is fired in the source system. 

In simple terms, Webhooks are "user-defined HTTP callbacks.” This callback is triggered after a specific event occurs in the source system, prompting the system to automatically send an HTTP request to the designated destination system.

## How Webhook works 

!["Image displaying how webhook works"](./images/using-webhook-with-node-red-how-webhook-works.png "Image displaying how webhook works"){data-zoomable}

- Event Initiator: This refers to the event specified to trigger the WebHook. Whenever this event occurs, the WebHook will be activated.

> - Webhook Server: The webhook server, is responsible for managing webhook configurations and endpoints. It listens for the specified event. When the event is detected, the webhook server automatically sends an HTTP POST request containing relevant data to the designated third-party application or service.

Reviewer: ```suggestion
- Webhook Server: The webhook server is responsible for managing webhook configurations and endpoints. It listens for the specified event. When the event is detected, the webhook server automatically sends an HTTP POST request containing relevant data to the designated third-party application or service.
```

- Data Reception by Third-Party Application: The third-party application will receive the data sent via the WebHook to the designated URL or listener provided during registration.
- Custom Action Execution: Upon receiving the POST request, specific actions can be performed.

## API Vs Webhook

It's common and understandable to get confused between APIs and webhooks, especially when you are learning about webhooks for the first time. However, comparing the two can help dispel these confusions.

| Aspect            | API                                                    | Webhook                                              |
|-------------------|--------------------------------------------------------|------------------------------------------------------|
| Direction         | Typically involves client-to-server communication.     | Typically involves server-to-server communication.   |

> | Initiation        | Requests are initiated by the client.                  | Requests are initiated by the server.                |

Reviewer: ```suggestion
| Initiation        | The client initiates requests.                  | The server initiates requests.                |
```

| Request Method    | Usually employs HTTP methods like GET, POST, etc.      | Typically uses the HTTP POST method.                 |
| Response          | Provides an immediate response upon request.           | Does not provide an immediate response; asynchronous.|
| Data Transfer     | Utilizes a pull model where the client fetches data.  | Operates on a push model where the server pushes data to the client.|
| Polling           | Requires periodic polling for updates.                 | No need for polling; receives updates directly.      |
| Payload           | The client specifies the payload in the request.       | The server defines the payload in the outgoing request. |
| Error Handling    | Typically includes error codes and messages.           | Errors are handled by retry mechanisms or manual intervention. |

### Example Scenario:

> Consider a manufacturing facility that utilizes temperature sensors to monitor temperature levels in critical areas. When the temperature surpasses predefined thresholds, it triggers a series of actions for maintenance and monitoring.

Reviewer: ```suggestion
Consider a manufacturing facility that utilizes temperature sensors to monitor temperature levels in critical areas. When the temperature exceeds predefined thresholds, it triggers a series of actions for maintenance and monitoring.
```

- Temperature Sensor (Server 1): These physical sensors are installed in the manufacturing plant to monitor temperature. They trigger webhook requests to Server 2 whenever abnormal temperature patterns are detected.

- Server 2 (Webhook Server): This server creates and hosts the webhook endpoint. It receives HTTP requests from the temperature sensors (Server 1) when abnormal temperatures are detected. The request contains temperature data. Server 2 then processes this data and sends a POST request with relevant information to Server 3.

- Maintenance System (Server 3): This system receives POST requests from Server 2 containing event-related data on a specific endpoint provided to Server 2. It then automatically schedules maintenance tasks based on the received information.

## Practical implementation

In this section, we will construct the practical implementation of the scenario described above. However, all three components or servers will be hosted on the same Node-RED instance in our example. Furthermore, we will simulate data rather than using real sensors by utilizing random number expressions.

### Setting Up a Webhook

> - Drag an http-in node onto the canvas. Configure the method as POST and set the path as "/test-webhook".

Reviewer: ```suggestion
- Drag an **http-in** node onto the canvas. Configure the method as POST and set the path as "/test-webhook".
```

!["Screenshot displaying webhook http-in nodes configuration"](./images/using-webhook-with-node-red-http-in-node-endpoint-for-receiving-data-from-server-2.png "Screenshot displaying webhook http-in nodes configuration"){data-zoomable}

> - Drag an http request node onto the canvas. Configure the method as POST and set the URL to `https://<your-instance-name>.flowfuse.cloud/schedule-maintenance`, replace <your-instance-name> with your actual name of the instance. "/schedule-maintenance" will be the endpoint for posting requests to the maintenance monitoring system provided by Server 3. 

Reviewer: ```suggestion
- Drag an **http request** node onto the canvas. Configure the method as POST and set the URL to `https://<your-instance-name>.flowfuse.cloud/schedule-maintenance`, replace <your-instance-name> with your actual name of the instance. "/schedule-maintenance" will be the endpoint for posting requests to the maintenance monitoring system provided by Server 3. 
```

!["Screenshot displaying http request nodes configuration for sending post request to server 3"](./images/using-webhook-with-node-red-request-node-sending-request-to-server3.png "Screenshot displaying http request nodes configuration for sending post request to server 3"){data-zoomable}

> - Drag an http response node onto the canvas and connect its input to the output of the http-in node. Also, connect an http request node's input to the same http in node's output.

Reviewer: ```suggestion
- Drag an **http response** node onto the canvas and connect its input to the output of the http-in node. Also, connect an http request node's input to the same http in node's output.
```

| Request Method    | Usually employs HTTP methods like GET, POST, etc.      | Typically uses the HTTP POST method.                 |
| Response          | Provides an immediate response upon request.           | Does not provide an immediate response; asynchronous.|
| Data Transfer     | Utilizes a pull model where the client fetches data.  | Operates on a push model where the server pushes data to the client.|
| Polling           | Requires periodic polling for updates.                 | No need for polling; receives updates directly.      |
| Payload           | The client specifies the payload in the request.       | The server defines the payload in the outgoing request. |
| Error Handling    | Typically includes error codes and messages.           | Errors are handled by retry mechanisms or manual intervention. |

### Example Scenario:

Consider a manufacturing facility that utilizes temperature sensors to monitor temperature levels in critical areas. When the temperature falls or exceeds predefined thresholds, it triggers a series of actions for maintenance and monitoring.

> !["Diagram explaining how component works in Webhook"](./images/using-webhook-with-node-red-diagram.png "Diagram explaining how component works in Webhook"){data-zoomable}

Author: @ZJvandeWeg I have also updated the diagram. Could you please check it and let me know if it is correct?

---

> title: Using Webhook with Node-RED

Reviewer 2: ```suggestion
title: Using Webhooks with Node-RED
```

> subtitle:  Learn to Integrate and Automate with Webhooks and Node-RED.

Reviewer 2: ```suggestion
subtitle: Learn to Integrate and Automate with Webhooks and Node-RED.
```

description: Learn how to seamlessly integrate webhooks into your Node-RED applications for automating tasks and enhancing communication. Get detailed, step-by-step instructions on setting up webhooks and practical implementation examples.
date: 2024-04-07
authors: ["sumit-shinde"]

> image:

Reviewer 2: Art request issue? Can we link it?

tags:
    - post
    - node-red
    - webhooks
    - event-driven
    - automation
---

> Have you found yourself needing to automate tasks based on specific events within your IoT applications? If so, utilizing webhooks presents an efficient solution for this challenge. Webhooks play an integral role in streamlining operations, facilitating real-time communication and seamless integration between disparate systems and devices. In this comprehensive guide, we cover everything you need to get started with webhooks, from basics to practical implementation. We provide detailed, step-by-step instructions on integrating webhooks within Node-RED applications.

Reviewer 2: ```suggestion
Have you found yourself needing to automate tasks based on specific events within your IoT applications? If so, utilizing webhooks presents an efficient solution for this challenge. Webhooks play an integral role in streamlining operations, facilitating real-time communication and seamless integration between disparate systems and devices. In this guide, we cover everything you need to get started with webhooks, from basics to practical implementation. We provide detailed, step-by-step instructions on integrating systems through webhooks using Node-RED.
```

<!--more-->

## What are the Webhooks?

> Webhooks are a mechanism that allows real-time communication between two different applications or services. They are essentially HTTP requests triggered by events in a source system and sent to a destination system, often with a payload of data. Webhooks are automated, meaning they are automatically sent out when their corresponding event is fired in the source system. 

Reviewer 2: "real-time" is isn't correct: https://en.wikipedia.org/wiki/Real-time_computing, almost no network protocol is real-time (there's a few though!). HTTP isn't designed to be.

```suggestion
Webhooks are a mechanism that allows fast communication between two different applications or services. They are essentially HTTP requests triggered by events in a source system and sent to a destination system, often with a payload of data. Webhooks are automated, meaning they are automatically sent out when their corresponding event is fired in the source system. 
```

## API:

- **Direction:**
  - Typically involves client-to-server communication.
- **Initiation:**
  - Requests are initiated by the client.
- **Request Method:**
  - Usually employs HTTP methods like GET, POST, etc.
- **Response:**
  - Provides an immediate response upon request.
- **Data Transfer:**
  - Utilizes a pull model where the client fetches data.
- **Polling:**
  - Requires periodic polling for updates.
- **Payload:**
  - The client specifies the payload in the request.
- **Error Handling:**
  - Typically includes error codes and messages.

## Webhook:

- **Direction:**
  - Typically involves server-to-server communication.
- **Initiation:**
  - Requests are initiated by the server.
- **Request Method:**
  - Typically uses the HTTP POST method.
- **Response:**
  - Does not provide an immediate response; asynchronous.
- **Data Transfer:**
  - Operates on a push model where the server pushes data to the client.
- **Polling:**
  - No need for polling; receives updates directly.
- **Payload:**
  - The server defines the payload in the outgoing request.
- **Error Handling:**

>   - Errors are handled by retry mechanisms or manual intervention.

Reviewer 2: Let's make a Markdown table, not two lists. Tables make comparing easier

### Example Scenario:

> For example, Consider a manufacturing facility that utilizes temperature sensors to monitor temperature levels in critical areas. When the temperature surpasses predefined thresholds, it triggers a series of actions for maintenance and monitoring.

Reviewer 2: ```suggestion
Consider a manufacturing facility that utilizes temperature sensors to monitor temperature levels in critical areas. When the temperature surpasses predefined thresholds, it triggers a series of actions for maintenance and monitoring.
```

> !["Diagram explaining how component works in"](./images/using-webhook-with-node-red-diagram.png "Role Based Access Control For Node-RED with FlowFuse"){data-zoomable}

Reviewer 2: Can you update the alt tag? this is a copy paste error I assume?

Author: Yes, it's a copy-paste error. I always begin by copying our previous blog post structure to prevent mistakes. I forgot to update this alt tag; I'll make sure to update it

- Temperature Sensor (Server 1): These physical sensors are installed in the manufacturing plant to monitor temperature. They trigger webhook requests to Server 2 whenever abnormal temperature patterns are detected.

> - Server 2 (Webhook Server): This server creates and hosts the webhook endpoint. It receives HTTP requests from the temperature sensors (Server 1) when abnormal temperatures are detected. The request contains temperature data. Server 2 then processes this data and sends a POST request with relevant information to Server 3.

Reviewer 2: Why have this intermediary server? Does this server process a lot of sensor readings?

Server 1 has limitations on how many messages it will send, so I believe you should explain why server 1 doesn't send the message straight to server 3.

Author: Okay ZJ, I'll provide an explanation for why Server 1 doesn't directly send messages to Server 3.

- Maintenance System (Server 3): This system receives POST requests from Server 2 containing event-related data on a specific endpoint provided to Server 2. It then automatically schedules maintenance tasks based on the received information.

## Practical implementation

> In this section, we will construct the practical implementation of the scenario described above. However, all three components or servers will be hosted on the same Node-RED instance in our example. Furthermore, we will simulate data rather than using real sensors by utilizing random number expressions.

Reviewer 2: I think this is cheating. I think you have a RPI, can you install the device agent on it and add a temperature sensor to it? You can expense both if you don't have it yet.

Than setup the device agent of Server 1, connect the sensor. Don't use Random

Author: Currently, I don't have an RPI or related equipment, but I'll purchase them as soon as possible. I'll make sure to update this blog and incorporate it into our next post.

Reviewer 2: Now you've got your RPI, please let me know when this post is ready :)

!["Screenshot displaying HTTP In node configuration for creating the POST request endpoint."](./images/using-webhook-with-node-red-http-in-node-endpoint-for-receiving-data-from-server-2.png "Screenshot displaying HTTP In node configuration for creating the POST request endpoint."){data-zoomable}

- Drag the function node onto Canvas and copy the below code in it.

```js
// Extract data from the incoming request
let data = msg.req.body;

// Retrieve or initialize scheduled maintenance data
let scheduledMaintenanceData = global.get('scheduledMaintenance') || [];

// Randomly assign maintenance task
let assignedTo = Math.random() < 0.5 ? "Bob Smith": "Alice Walker";

// Create a recent maintenance schedule record
let maintenanceScheduleRecentData = {
    "ocured_at": new Date().toISOString(), // Current timestamp
    ...data, // Include incoming data
    "assigned_to": assignedTo // Assign to a worker
};

// Add recent maintenance data to records
scheduledMaintenanceData.push(maintenanceScheduleRecentData);

// Update scheduled maintenance data to global context
global.set('scheduledMaintenance', scheduledMaintenanceData);

return msg;

> ```

Reviewer 2: I feel half this code block should be a change node, not a function node.
